handler/src/test/java/io/netty/handler/stream/ChunkedWriteHandlerTest.java

```diff
@@ -227,6 +228,207 @@ public long progress() {
         assertNull(ch.readOutbound());
     }
 
+    @Test
+    public void testWriteFailureChunkedStream() throws IOException {
+        checkFirstFailed(new ChunkedStream(new ByteArrayInputStream(BYTES)));
+    }
+
+    @Test
+    public void testWriteFailureChunkedNioStream() throws IOException {
+        checkFirstFailed(new ChunkedNioStream(Channels.newChannel(new ByteArrayInputStream(BYTES))));
+    }
+
+    @Test
+    public void testWriteFailureChunkedFile() throws IOException {
+        checkFirstFailed(new ChunkedFile(TMP));
+    }
+
+    @Test
+    public void testWriteFailureChunkedNioFile() throws IOException {
+        checkFirstFailed(new ChunkedNioFile(TMP));
+    }
+
+    @Test
+    public void testWriteFailureUnchunkedData() throws IOException {
+        checkFirstFailed(Unpooled.wrappedBuffer(BYTES));
+    }
+
+    @Test
+    public void testSkipAfterFailedChunkedStream() throws IOException {
+        checkSkipFailed(new ChunkedStream(new ByteArrayInputStream(BYTES)),
+                        new ChunkedStream(new ByteArrayInputStream(BYTES)));
+    }
+
+    @Test
+    public void testSkipAfterFailedChunkedNioStream() throws IOException {
+        checkSkipFailed(new ChunkedNioStream(Channels.newChannel(new ByteArrayInputStream(BYTES))),
+                        new ChunkedNioStream(Channels.newChannel(new ByteArrayInputStream(BYTES))));
+    }
+
+    @Test
+    public void testSkipAfterFailedChunkedFile() throws IOException {
+        checkSkipFailed(new ChunkedFile(TMP), new ChunkedFile(TMP));
+    }
+
+    @Test
+    public void testSkipAfterFailedChunkedNioFile() throws IOException {
+        checkSkipFailed(new ChunkedNioFile(TMP), new ChunkedFile(TMP));
+    }
+
+    // See https://github.com/netty/netty/issues/8700.
+    @Test
+    public void testFailureWhenLastChunkFailed() throws IOException {
+        ChannelOutboundHandlerAdapter failLast = new ChannelOutboundHandlerAdapter() {
+            private int passedWrites;
+
+            @Override
+            public void write(ChannelHandlerContext ctx, Object msg, ChannelPromise promise) {
+                if (++this.passedWrites < 4) {
+                    ctx.write(msg, promise);
+                } else {
+                    promise.tryFailure(new RuntimeException());
```

Comment:
You also need to call `ReferenceCountUtil.release(msg)` before failing the promise.